### PR TITLE
Separate model object in non-linear extension

### PR DESCRIPTION
**Pull request by ajreynol**
This separates all functionality in the nonlinear extension regarding computing model values and establishing SAT answers to its own file.

There are no major modifications to the overall behavior of the code, although some interfaces were refactored to be clearer.

This is work towards correcting issues in the models generated by the nonlinear solver (#2153).

### CODE REVIEW

File: src/theory/arith/nl_model.h [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+   *
+   * For example, if M( a ) = 2, M( b ) = 3, M( a*b ) = 5, i.e. the variable
+   * for a*b has been assigned a value 5 by the linear solver, then :
+   *
+   *   computeModelValue( a*b, true ) =
+   *   computeModelValue( a, true )*computeModelValue( b, true ) = 2*3 = 6
+   * whereas:
+   *   computeModelValue( a*b, false ) = 5
+   */
+  Node computeConcreteModelValue(Node n);
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_model.h]: Maybe we should put the function body here or inline the function.

File: src/theory/arith/nl_model.h [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+  Node computeAbstractModelValue(Node n);
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_model.h]: ditto

File: src/theory/arith/nl_model.h [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+  bool checkModel(const std::vector<Node>& assertions,
+                  const std::vector<Node>& false_asserts,
+                  unsigned d,
+                  std::vector<Node>& lemmas,
+                  std::vector<Node>& gs);
+  /**
+   * Set that we have used an approximation during this check. This flag is
+   * reset on a call to resetCheck. It is set when we use reasoning that
+   * is limited by a degree of precision we are using. In other words, if we
+   * used an approximation, then we maybe could still establish or lemma or
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_model.h]: typo: establish a lemma

File: src/theory/arith/nl_model.h [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+   * determine the input in SAT if we increased our precision.
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_model.h]: typo: input is SAT

File: src/theory/arith/nl_model.cpp [hunk 5]
```diff
@@ -0,0 +1,1295 @@
...
+}
+
+bool NlModel::isRefineableTfFun(Node tf)
+{
+  Assert(tf.getKind() == SINE || tf.getKind() == EXPONENTIAL);
+  if (tf.getKind() == SINE)
+  {
+    // we do not consider e.g. sin( -1*x ), since considering sin( x ) will
+    // have the same effect
+    if (!tf[0].isVar())
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_model.cpp]: what about `sin(x+y)`

File: src/theory/arith/nl_model.h [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+  //---------------------------check model
+  /** solve equality simple
+   *
+   * This method is used during checkModel(...). It takes as input an
+   * equality eq. If it returns true, then eq is correct-by-construction based
+   * on the information stored in our model representation (see
+   * d_check_model_vars, d_check_model_subs, d_check_model_bounds), and eq
+   * is added to d_check_model_solved.
+   */
+  bool solveEqualitySimple(Node eq, unsigned d, std::vector<Node>& lemmas);
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_model.h]: does it work for any nonlinear equality or just univariate quadratic equations?

File: src/theory/arith/nl_model.h [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+                  std::vector<Node>& lemmas,
+                  std::vector<Node>& gs);
+  /**
+   * Set that we have used an approximation during this check. This flag is
+   * reset on a call to resetCheck. It is set when we use reasoning that
+   * is limited by a degree of precision we are using. In other words, if we
+   * used an approximation, then we maybe could still establish or lemma or
+   * determine the input in SAT if we increased our precision.
+   */
+  void setUsedApproximate();
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_model.h]: maybe put the function body here or inline the function

File: src/theory/arith/nl_model.h [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+  /** Did we use an approximation during this check? */
+  bool usedApproximate() const;
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_model.h]: ditto

File: src/theory/arith/nl_model.cpp [hunk 9]
```diff
@@ -0,0 +1,1295 @@
...
+  Trace("nl-ext-cms") << "simple solve equality " << seq << "..." << std::endl;
+  Assert(seq.getKind() == EQUAL);
+  std::map<Node, Node> msum;
+  if (!ArithMSum::getMonomialSumLit(seq, msum))
+  {
+    Trace("nl-ext-cms") << "...fail, could not determine monomial sum."
+                        << std::endl;
+    return false;
+  }
+  bool is_valid = true;
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_model.cpp]: what happens if the degree of monomial is greater than 2?

File: src/theory/arith/nl_model.h (hunk 8 above)
Comment (ajreynol) [reply src/theory/arith/nl_model.h]: see above

**Pull request merged**